iotlab_controller/riot.py

```diff
@@ -41,23 +41,19 @@ def path(self):
         else:
             return self.flashfile
 
-    def build(self, build_env=None, threads=1):
+    def run(self, build_env, cmd):
         env = self.env.copy()
         if build_env is not None:
             env.update(build_env)
         try:
-            subprocess.run(["make", "-j", str(threads), "-C",
-                           self.application_path, "all"],
-                           env=env, check=True)
+            subprocess.run(cmd, env=env, check=True)
         except subprocess.CalledProcessError as e:
             raise firmware.FirmwareBuildError(e)
 
-    def clean(self, build_env=None):
-        env = self.env.copy()
-        if build_env is not None:
-            env.update(build_env)
-        try:
-            subprocess.run(["make", "-C", self.application_path, "clean"],
-                           env=env, check=True)
-        except subprocess.CalledProcessError as e:
-            raise firmware.FirmwareBuildError(e)
+    def build(self, build_env=None, threads=1):
+        self.run(build_env, ["make", "-j", str(threads), "-C",
+                             self.application_path, "all"])
+
+    def clean(self, build_env=None, distclean=False):
+        op = "distclean" if distclean else "clean"
```

Comment:
`distclean` does not imply `clean` (see https://github.com/RIOT-OS/RIOT/blob/55cc34fd174c241df69e9992024ed0590f2255f6/Makefile.include#L653-L655 vs. https://github.com/RIOT-OS/RIOT/blob/55cc34fd174c241df69e9992024ed0590f2255f6/Makefile.include#L640-L643), so this should rather be:

```suggestion
        op = ["distclean"] if distclean else ["distclean", "clean"]
```

(and the usage of op below adapted to `["make", "-C", self.application_path] + op`). However, I don't see why `distclean` should be within `clean`. Rather I would prefer a separate method.

Reply:
ok, had it separately in my first iteration, as in my testing a `distclean` always removed my complste `app/bin` folder. But as you state, going back to that should be the saver way -> will fix.

Reply:
I've seen the removal of `app/bin` in https://github.com/RIOT-OS/RIOT/blob/55cc34fd174c241df69e9992024ed0590f2255f6/Makefile.include#L653-L655, however I'm not quite sure why it is their... given RIOT's current directory structure (having packages checked out within `app/bin` the first line seems to be redundant then).